glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,375 @@
+                        .filter_map(|line| {
+                            println!("{}", line);
+                            serde_json::from_str::<JsonValue>(&line).ok()
+                        })
+                        .filter_map(|json| RecordedEventData::deserialize(&json))
+                        .collect(),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    fn send_all_events(&self, glean: &Glean) {
+        let store_names = {
+            let db = self.event_stores.read().unwrap();
+            db.keys().cloned().collect::<Vec<String>>()
+        };
+
+        for store_name in store_names {
+            if let Err(err) = glean.send_ping_by_name(&store_name, false) {
+                log::error!(
+                    "Error flushing existing events to the {} ping: {}",
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Record an event in the desired stores.
+    ///
+    /// # Arguments
+    ///
+    /// - `glean`: The Glean instance.
+    /// - `meta`: The metadata about the event metric. Used to get the category,
+    ///   name and stores for the metric.
+    /// - `timestamp`: The timestamp of the event, in nanoseconds. Must use a
+    ///   monotonically increasing timer (this value is obtained on the
+    ///   platform-specific side).
+    /// - `extra`: Extra data values, mapping strings to strings.
+    pub fn record(
+        &self,
+        glean: &Glean,
+        meta: &CommonMetricData,
+        timestamp: u64,
+        extra: Option<HashMap<String, String>>,
+    ) {
+        let event = RecordedEventData {
+            timestamp,
+            category: meta.category.to_string(),
+            name: meta.name.to_string(),
+            extra,
+        };
+        let event_json = event.serialize().to_string();
+        let mut stores_to_send: Vec<&str> = Vec::new();
+
+        {
+            let mut db = self.event_stores.write().unwrap();
+            for store_name in meta.send_in_pings.iter() {
+                let store = db
+                    .entry(store_name.to_string())
+                    .or_insert_with(|| Vec::new());
+                store.push(event.clone());
+                self.write_event_to_disk(store_name, &event_json);
+                if store.len() == glean.get_max_events() {
+                    stores_to_send.push(&store_name);
+                }
+            }
+        }
+
+        for store_name in stores_to_send {
+            if let Err(err) = glean.send_ping_by_name(store_name, false) {
+                log::error!(
+                    "Got more than {} events, but could not send {} ping: {}",
+                    glean.get_max_events(),
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Writes an event to a single store on disk.
+    ///
+    /// This assumes that the write lock in `self.event_stores` is already held.
```

Comment:
Why is this relevant? If it is, we should build this function in a way that this is required.

Reply:
We don't want multiple threads writing to files at the same time.  On glean-ac, this was handled using an explicit sync variable *just* for the filesystem operations.  In this implementation, we always hold a lock on the in-memory data structure whenever we do file system operations anyway.  We could add a RwLock just for the on-disk stuff for clarity, but it would be redundant.

Reply:
I've updated this to have an explicit lock for file operations, separate from the in-memory data structure's lock.